Expand and update docs on monorepo package publishing

jsnajdr opened the pull request:
Today with @tyxla we did a Zoom call with a walkthrough of the NPM publishing process. This documentation update in one of the outcomes. It describes the process and all the gotchas and options in much more detail and hopefully it will be useful.

jsnajdr committed:
Expand and update docs on monorepo package publishing

tyxla commented on docs/monorepo.md:
```diff
@@ -133,22 +133,70 @@ Note that if you're building with Webpack, you may need to turn off [`resolve.sy
+
+The rest of the workflow is exactly the same as in the `from-package` case.
+
+### Getting NPM permissions to publish in the `@automattic` scope
```
Should this be moved above as it's a prerequisite for being able to publish?

tyxla commented on docs/monorepo.md:
```diff
@@ -133,22 +133,70 @@ Note that if you're building with Webpack, you may need to turn off [`resolve.sy
+
+Make sure that the `CHANGELOG.md` document contains up-to-date information, with the `next` heading replaced with the version number that you are about to publish.
+
+Create PRs with the necessary changes and merge them to `master` before publishing. Lerna will add a `gitHead` field to each published package's `package.json`. That field contains the hash of the Git commit that the package was published from. It's better if this commit hash is a permanent one from the `master` branch, rather than an ephemeral commit from a local branch.
```
👍

tyxla reviewed (approved):
This looks awesome, thank you @jsnajdr! Just left a minor optional suggestion, but feel free to 🚢 either way!

sirreal commented on docs/monorepo.md:
````diff
@@ -133,22 +133,70 @@ Note that if you're building with Webpack, you may need to turn off [`resolve.sy
+
+```
+NPM_CONFIG_OTP=[YOUR_OTP_CODE] npx lerna publish --dist-tag next from-package --yes
+```
````
I think we can remove the OTP cycle time and `NPM_CONFIG_OTP=…` bits here. `lerna` now asks for OTP at the right time.

jsnajdr replied:
Alright. I can confirm that Lerna/NPM asks interactively if you don't set the environment variable.

sirreal commented on docs/monorepo.md:
```diff
@@ -133,22 +133,70 @@ Note that if you're building with Webpack, you may need to turn off [`resolve.sy
+
+Now `lerna publish from-git` will offer to publish only the packages that have a matching Git tag on the current `HEAD` revision.
+
+The rest of the workflow is exactly the same as in the `from-package` case.
```
Do we want to suggest `from-git`? It seemed very problematic when we tried to use it early on. Maybe we should suggest adding `private:true` to `package.json` temporarily.

tyxla replied:
Seems like it's the most straightforward way to publish a single package. The `private:true` one seems even more like a hack 😉 What was problematic about it before?

jsnajdr replied:
@sirreal I use `from-git` every time I'm publishing, haven't run into any issues so far, and I can recommend this workflow 🙂 If there are any risks, let's describe them.

jsnajdr committed:
Updated docs after review feedback

jsnajdr commented:
Pushed an update that addresses review feedback:
- put the section about NPM accounts and the `automattic` org before we start describing `lerna publish`
- describe the interactive flow (with fewer CLI) options by default. Describe the non-interactive options in a premium section
- describe the `--dist-tag next` option in a dedicated section. Keep the main line simple.

sirreal reviewed (approved):
Thanks for updating these!
